test/types/client_bulk_write.test-d.ts

```diff
@@ -0,0 +1,258 @@
+    update: {}
+  },
+  {
+    namespace: 'db.up',
+    name: 'updateMany',
+    filter: { name: 'bob', published: 2 },
+    update: {}
+  },
+  { namespace: 'db.random', name: 'deleteOne', filter: {} },
+  { namespace: 'db.namespace', name: 'deleteMany', filter: {} }
+]);
+
+// Operation names are still type checked when there is no schema
+expectError<ClientBulkWriteModel>({
+  namespace: 'db.author',
+  name: 'insertLots', // Not an operation we support
+  document: { name: 'bob', published: 2 }
+});
+
+type MongoDBSchemas = {
+  'db.books': Book;
+  'db.authors': Author;
+  'db.stores': Store;
+};
+
+expectError<ClientBulkWriteModel<MongoDBSchemas>>({
+  namespace: 'db.author', // Unknown namespace! a typo!
+  name: 'insertOne',
+  document: { name: 'bob', published: 2 }
+});
+
+expectError<ClientBulkWriteModel<MongoDBSchemas>>({
+  namespace: 'db.authors',
+  name: 'insertOne',
+  document: { name: 'bob', published: '' } // Incorrect type for known field
+});
+
+expectError<ClientBulkWriteModel<MongoDBSchemas>>({
+  namespace: 'db.authors',
+  name: 'insertOne',
+  document: { name: 'bob', publish: 2 } // unknown field! typo!
+});
+
+// Defined schemas - all correct use
+client.bulkWrite<MongoDBSchemas>([
+  {
+    namespace: 'db.authors',
+    name: 'insertOne',
+    document: { name: 'bob', published: 2 }
+  },
+  {
+    namespace: 'db.authors',
+    name: 'replaceOne',
+    filter: { name: 'bob' },
+    replacement: { name: 'ann', published: 2 }
+  },
+  {
+    namespace: 'db.authors',
+    name: 'updateOne',
+    filter: { name: 'bob', published: 2 },
+    update: {}
+  },
+  {
+    namespace: 'db.authors',
+    name: 'updateMany',
+    filter: { name: 'bob', published: 2 },
+    update: {}
+  },
+  { namespace: 'db.authors', name: 'deleteOne', filter: {} },
+  { namespace: 'db.authors', name: 'deleteMany', filter: {} }
+]);
+
+// Defined schemas - incorrect use
+expectError(
+  client.bulkWrite<MongoDBSchemas>([
+    {
+      namespace: 'db.authors',
+      name: 'insertOne',
+      document: { name: 'bob', published: '' } // wrong type
+    },
+    {
+      namespace: 'db.authors',
+      name: 'replaceOne',
+      filter: { name: 'bob' },
+      replacement: { name: 'ann', publish: 2 } // key typo
+    },
+    {
+      namespace: 'db.blah', // unknown namespace
+      name: 'updateOne',
+      filter: { name: 'bob', published: 2 },
+      update: {}
+    },
+    {
+      namespace: 'db.authors',
+      name: 'updateManyy', // unknown operation
+      filter: { name: 'bob', published: 2 },
+      update: {}
+    },
+    { namespace: 'db.authors', name: 'deleteOne', filter: {} },
+    { namespace: 'db.authors', name: 'deleteMany', filter: {} }
+  ])
+);
+
+type MongoDBSchemasWithCalculations = {
+  // River Books uses star ratings
+  [key: `river-books.${string}`]: Book & { fiveStarRatings: number };
+  // Ocean literature uses thumbs up for ratings
+  [key: `ocean-literature.${string}`]: Book & { thumbsUp: number };
+};
```

Comment:
What is the use case for dynamically-created namespaces?

Reply:
This code snippet is kinda the use case, it is common to create namespaces that reuse the same type so you can avoid actually writing them all out and just define the prefix

Reply:
Usually you'd just use the [polymorphic pattern](https://www.mongodb.com/developer/products/mongodb/polymorphic-pattern/) for something like this.

Sharing schemas is still possible, just orthogonal to dynamically computed namespaces:

```typescript
type MongoDBSchemasWithCalculations = {
  // River Books uses star ratings
  [key: `river-books.b`]: Book & { fiveStarRatings: number };
  // Ocean literature uses thumbs up for ratings
  [key: `ocean-literature.b`]: Book & { thumbsUp: number };
};
```

Reply:
Yea I guess I gave the wrong impression with "withCalculations" I meant type calculations not dynamic at runtime necessarily. 

Here's a more simple case, not related to data design per se:

```ts
type MongoDBSchemas = { [key: `bookstores.${string}`]: Book };
```

Now all collections in the bookstores db are using the Book schema. I don't have to write out all my collections this way.